Add features - add random request ID generation. uk27 random protos 2.19.1 and uk25

Furtif opened the pull request:
### Support of 0.79.4

Furtif committed:
add features

Furtif committed:
Merge branch 'add_missed' into new

Furtif committed:
fix const

Furtif committed:
typo comments

Furtif committed:
missed inbox

Furtif committed:
sync data

Furtif committed:
bug fix

Furtif committed:
typo code

Furtif committed:
update pogoprotos

AeonLucid commented on POGOLib.Official/Net/RpcEncryption.cs:
```diff
@@ -190,8 +190,11 @@ private List<LocationFix> BuildLocationFixes(RequestEnvelope requestEnvelope, lo
-            signature.LocationHash1 = (int) hashData.LocationAuthHash;
-            signature.LocationHash2 = (int) hashData.LocationHash;
+            if (hashData == null)
+                return null;
```
Maybe better to throw an exception because `null` is not expected here.

Furtif replied:
Aff this please :-)

Furtif replied:
Add*

AeonLucid commented on POGOLib.Official/Util/Hash/PokeHashHasher.cs:
```diff
@@ -218,9 +230,13 @@ private Task<HttpResponseMessage> PerformRequest(HttpContent requestContent)
+
+                    if (response.Headers.TryGetValues("X-MaxRequestCount", out maxRequestsValue) &&
+                        response.Headers.TryGetValues("X-RateRequestsRemaining", out requestsRemainingValue) &&
+                        response.Headers.TryGetValues("X-RatePeriodEnd", out ratePeriodEndValue))
```
Why?

Furtif replied:
This I do not know well I saw this in another github

AeonLucid commented on POGOLib.Official/Net/RpcClient.cs:
```diff
@@ -63,6 +65,9 @@ public class RpcClient : IDisposable
         private readonly Semaphore _rpcQueueMutex = new Semaphore(1, 1);
 
+        public event EventHandler<GetHatchedEggsResponse> HatchedEggsReceived;
+        public event EventHandler<CheckAwardedBadgesResponse> CheckAwardedBadgesReceived;
```
These shouldn't be in `RpcClient`. This class is meant for networking.

Furtif replied:
Change this please :-)

Furtif replied:
it is needed for https://github.com/AeonLucid/POGOLib/pull/97/files#diff-f9807ab8e1379a13294d996b27e4d204R749

AeonLucid commented on POGOLib.Official/Net/RpcClient.cs:
```diff
@@ -231,7 +244,16 @@ private void IncrementRequestCount()
+            {
+                var cell = _session.Map.Cells.FirstOrDefault(x => x.S2CellId == cellId);
+
+                // TODO: Using the cells' currenttimestamp gives a delta, which is probably better, but we need to merge the result with the existing ones
```
This is what I did before but people complained that it was wrong..

AeonLucid commented on POGOLib.Official/Net/RpcClient.cs:
```diff
@@ -441,7 +463,11 @@ private IEnumerable<Request> GetDefaultRequests()
 
-            requestEnvelope.PlatformRequests.Add(await _rpcEncryption.GenerateSignatureAsync(requestEnvelope));
+            PlatformRequest platformRequest = await _rpcEncryption.GenerateSignatureAsync(requestEnvelope);
+            if (platformRequest != null)
```
Why? Request must have a signature.

AeonLucid commented on POGOLib.Official/Net/RpcClient.cs:
```diff
@@ -605,7 +631,11 @@ private Task<ByteString> SendRemoteProcedureCall(RequestEnvelope requestEnvelope
 
-                                requestEnvelope.PlatformRequests.Insert(0, await _rpcEncryption.GenerateSignatureAsync(requestEnvelope));
+                                PlatformRequest platformRequest = await _rpcEncryption.GenerateSignatureAsync(requestEnvelope);
+                                if (platformRequest != null)
```
Why? Request must have a signature.

Furtif commented:
I think it's best if I let you see that it will not make you screw up

Furtif committed:
fixes:
https://github.com/AeonLucid/POGOLib/pull/97#discussion_r130258004
https://github.com/AeonLucid/POGOLib/pull/97#discussion_r130258026

Furtif committed:
fixes https://github.com/AeonLucid/POGOLib/pull/97#discussion_r130258004

Furtif committed:
Fix value

Furtif committed:
fixe
https://github.com/AeonLucid/POGOLib/pull/97#discussion_r130257774 https://github.com/AeonLucid/POGOLib/pull/97#discussion_r130257804

Furtif committed:
update login agent

Furtif committed:
typo code

Furtif committed:
change inbox methode

Furtif committed:
hello appveyor

Furtif committed:
random request ID generation.

For testing and cross-API consistency purposes. Still needs proper testing.

The Node.JS API maintainers claim to have found a fix related to shadowbans in the random request ID generation.

For testing purposes, I've implemented the same method in pgoapi. I've also removed old hardcoded values and changed class variables to be instance variables.

Furtif committed:
Add uk27 and uk25 0.75.1

Furtif commented:
@AeonLucid   Merge https://github.com/AeonLucid/POGOProtos/pull/309 update [POGOProtos.NetStandard1](https://github.com/AeonLucid/POGOProtos.NetStandard1)

Furtif committed:
lasted POGOProtos.NetStandard1

Furtif committed:
Add iphone 7 iOS 11.1.0 settings

Furtif committed:
ops... missing ,

Furtif committed:
add iphone 6s plus ios 10.3.3

Furtif committed:
change ordre comments...

Furtif committed:
Update nuget protos pack

Furtif committed:
according uk25 to 0.77.1

Furtif committed:
lasted protos

Furtif committed:
0.79.3

PeterDM65 commented:
Hi,

After having problems with a previous version of PogoLib since last week (before it worked just fine), i was checking out the latest version of Furtif's developers branch of PogoLib.

Running the following demo program : "POGOLib.Official.Demo.ConsoleApp" after setting the values for "pokeHashAuthKey", "loginProviderStr" (=google), "usernameStr" and "passwordStr", I got the following loggings:


2017-10-21 13:20:09.0579  INFO Booting up.
2017-10-21 13:20:09.1109  INFO Type 'q', 'quit' or 'exit' to exit.
2017-10-21 13:20:10.8024 DEBUG Authenticated through Google.
2017-10-21 13:20:10.8024 DEBUG Session state was set to Stopped.
2017-10-21 13:20:10.8821 DEBUG CheckHasherVersion
2017-10-21 13:20:11.1914  INFO PogoVersion:0.79.3 - Hasher.PokemonVerson:0.79.3 
2017-10-21 13:20:11.1914 DEBUG Session state was set to Started.
2017-10-21 13:20:12.0274 DEBUG Sending RPC Request: 'GetPlayer, CheckChallenge' 
2017-10-21 13:20:12.0274 DEBUG => Platform Request: 'SendEncryptedSignature, UnknownPtr8'
2017-10-21 13:20:12.6126 DEBUG Sending RPC Request: 'GetPlayer, CheckChallenge' 
2017-10-21 13:20:12.6126 DEBUG => Platform Request: 'SendEncryptedSignature, UnknownPtr8'
2017-10-21 13:20:12.8869 DEBUG Received a new AuthTicket from Pokemon!
2017-10-21 13:20:14.0028 DEBUG Sending RPC Request: 'GetMapObjects, CheckChallenge, GetHatchedEggs, GetInventory, CheckAwardedBadges, DownloadSettings, GetInbox'
2017-10-21 13:20:14.0028 DEBUG => Platform Request: 'SendEncryptedSignature, UnknownPtr8'
2017-10-21 13:20:14.1558  INFO Unknown status code: InvalidRequest
2017-10-21 13:20:14.1811 ERROR SendRemoteProcedureCall exception: System.Exception: There were 0 responses.
   at POGOLib.Official.Net.RpcClient.HandleResponseEnvelope(RequestEnvelope requestEnvelope, ResponseEnvelope responseEnvelope) in D:\Gebruikers\Peter\Downloads\POGOLib-ec3b6129adb1ecf5a37d5a45917e4eedd382f642\POGOLib-ec3b6129adb1ecf5a37d5a45917e4eedd382f642\POGOLib.Official\Net\RpcClient.cs:line 688
   at POGOLib.Official.Net.RpcClient.<PerformRemoteProcedureCallAsync>d__41.MoveNext() in D:\Gebruikers\Peter\Downloads\POGOLib-ec3b6129adb1ecf5a37d5a45917e4eedd382f642\POGOLib-ec3b6129adb1ecf5a37d5a45917e4eedd382f642\POGOLib.Official\Net\RpcClient.cs:line 663
2017-10-21 13:20:14.2463 ERROR Map refresh failed: System.NullReferenceException: response
   at POGOLib.Official.Net.RpcClient.<RefreshMapObjectsAsync>d__33.MoveNext() in D:\Gebruikers\Peter\Downloads\POGOLib-ec3b6129adb1ecf5a37d5a45917e4eedd382f642\POGOLib-ec3b6129adb1ecf5a37d5a45917e4eedd382f642\POGOLib.Official\Net\RpcClient.cs:line 301
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at POGOLib.Official.Pokemon.HeartbeatDispatcher.<Dispatch>d__7.MoveNext() in D:\Gebruikers\Peter\Downloads\POGOLib-ec3b6129adb1ecf5a37d5a45917e4eedd382f642\POGOLib-ec3b6129adb1ecf5a37d5a45917e4eedd382f642\POGOLib.Official\Pokemon\HeartbeatDispatcher.cs:line 124
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at POGOLib.Official.Pokemon.HeartbeatDispatcher.<CheckDispatch>d__4.MoveNext() in D:\Gebruikers\Peter\Downloads\POGOLib-ec3b6129adb1ecf5a37d5a45917e4eedd382f642\POGOLib-ec3b6129adb1ecf5a37d5a45917e4eedd382f642\POGOLib.Official\Pokemon\HeartbeatDispatcher.cs:line 73
2017-10-21 13:20:14.2463  INFO No fort found nearby.

And this is exactly the same problem with the previous version.
What went wrong ? 
The RPC Request "GetPlayer" is oké, I see my player stats. But it crashes afterwards ...

Furtif commented:
Pogolib into nuguets is outed ... Check source of this current pr or try m’y core branch

PeterDM65 commented:
Thnx for the response,
but ... 
I downloaded as zip your source code @ "https://github.com/Furtif/POGOLib/tree/core"
and run again the demo programm "POGOLib.Core.Demo.ConsoleApp" (after changing the hashkey, username, password of course).
And got the same logging (Unknown status code : InvalidRequest)

2017-10-21 16:34:45.8713  INFO Booting up.
2017-10-21 16:34:45.9252  INFO Type 'q', 'quit' or 'exit' to exit.
2017-10-21 16:34:58.0824 DEBUG Authenticated through Google.
2017-10-21 16:34:58.0824 DEBUG Session state was set to Stopped.
2017-10-21 16:35:08.8814 DEBUG Session state was set to Started.
2017-10-21 16:35:10.0218 DEBUG Sending RPC Request: 'GetPlayer, CheckChallenge' 
2017-10-21 16:35:10.0218 DEBUG => Platform Request: 'SendEncryptedSignature, UnknownPtr8'
2017-10-21 16:35:10.5968 DEBUG Sending RPC Request: 'GetPlayer, CheckChallenge' 
2017-10-21 16:35:10.5968 DEBUG => Platform Request: 'SendEncryptedSignature, UnknownPtr8'
2017-10-21 16:35:10.8348 DEBUG Received a new AuthTicket from Pokemon!
2017-10-21 16:35:11.8351 DEBUG Sending RPC Request: 'GetMapObjects, CheckChallenge, GetHatchedEggs, GetInventory, CheckAwardedBadges, DownloadSettings, GetInbox'
2017-10-21 16:35:11.8351 DEBUG => Platform Request: 'SendEncryptedSignature, UnknownPtr8'
2017-10-21 16:35:12.0620  INFO Unknown status code: InvalidRequest
2017-10-21 16:35:12.0883 ERROR SendRemoteProcedureCall exception: System.Exception: There were 0 responses.
   at POGOLib.Official.Net.RpcClient.HandleResponseEnvelope(RequestEnvelope requestEnvelope, ResponseEnvelope responseEnvelope) in D:\Gebruikers\Peter\Downloads\POGOLib-core\POGOLib-core\POGOLib.Core\Net\RpcClient.cs:line 687
   at POGOLib.Official.Net.RpcClient.<PerformRemoteProcedureCallAsync>d__41.MoveNext() in D:\Gebruikers\Peter\Downloads\POGOLib-core\POGOLib-core\POGOLib.Core\Net\RpcClient.cs:line 662
2017-10-21 16:35:12.1538 ERROR Map refresh failed: System.NullReferenceException: response
   at POGOLib.Official.Net.RpcClient.<RefreshMapObjectsAsync>d__33.MoveNext() in D:\Gebruikers\Peter\Downloads\POGOLib-core\POGOLib-core\POGOLib.Core\Net\RpcClient.cs:line 301
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at POGOLib.Official.Pokemon.HeartbeatDispatcher.<Dispatch>d__7.MoveNext() in D:\Gebruikers\Peter\Downloads\POGOLib-core\POGOLib-core\POGOLib.Core\Pokemon\HeartbeatDispatcher.cs:line 124
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at POGOLib.Official.Pokemon.HeartbeatDispatcher.<CheckDispatch>d__4.MoveNext() in D:\Gebruikers\Peter\Downloads\POGOLib-core\POGOLib-core\POGOLib.Core\Pokemon\HeartbeatDispatcher.cs:line 73
2017-10-21 16:35:35.8067  INFO No fort found nearby.

Furtif commented:
you uses a valid key hash ??

PeterDM65 commented:
Yes,
It's a new one, and the request for "GetPlayer" seems to work ...
This key worked till last monday

Furtif commented:
ok check procedures

- GYM_GET_INFO replaces GET_GYM_DETAILS (*need edition)
- GET_HOLO_INVENTORY  replaces GET_INVENTORY  (Edited https://github.com/AeonLucid/POGOLib/pull/97/commits/6e0b793df22d81b79f786c37dc9b9fd9c6685b68)
- GYM_START_SESSION replaces START_GYM_BATTLE  (*need edition)
- GYM_BATTLE_ATTACK replaces ATTACK_GYM (*need edition)

```*Maybe```

PeterDM65 commented:
Yes ... that seems to do it

I replaced "RequestType.GetInventory" by "RequestType.GetHoloInventory", and the demo program 
worked, at following places

Find all "GetInventory", Subfolders, Find Results 1, Entire Solution, ""
  D:\Gebruikers\Peter\Downloads\POGOLib-core\POGOLib-core\POGOLib.Core\Net\RpcClient.cs(59):            RequestType.GetInventory,
  D:\Gebruikers\Peter\Downloads\POGOLib-core\POGOLib-core\POGOLib.Core\Net\RpcClient.cs(363):                    RequestType = RequestType.GetInventory,
  D:\Gebruikers\Peter\Downloads\POGOLib-core\POGOLib-core\POGOLib.Core\Net\RpcClient.cs(364):                    RequestMessage = new GetInventoryMessage
  D:\Gebruikers\Peter\Downloads\POGOLib-core\POGOLib-core\POGOLib.Core\Net\RpcClient.cs(763):                    case RequestType.GetInventory: // Get_Inventory
  D:\Gebruikers\Peter\Downloads\POGOLib-core\POGOLib-core\POGOLib.Core\Net\RpcClient.cs(764):                        var inventory = GetInventoryResponse.Parser.ParseFrom(bytes);
  C:\Users\Peter\AppData\Local\JetBrains\Shared\v09\DecompilerCache\decompiler\682D9FC3-A62C-4079-BDE9-7C5EE24C0FCB\c3\c0a60562\RequestType.cs(102):    [OriginalName("GET_INVENTORY")] GetInventory = 5005,
  Matching lines: 6    Matching files: 2    Total files searched: 53

Thanx very much !!!!!

Furtif committed:
fix inventory

Furtif commented:
@PeterDM65 fixed get inventory redownload sources.

Furtif committed:
update protos

Furtif committed:
1.49.4

Furtif committed:
minor fixes

PeterDM65 commented:
@Furtif: I downloaded your latest version of the pogolib at https://github.com/Furtif/POGOLib (core branch). When running the demo program "POGOLib.Core.Demo.ConsoleApp" I got the following errors:
2017-11-10 23:14:29.5564  INFO Booting up.
2017-11-10 23:14:29.6058  INFO Type 'q', 'quit' or 'exit' to exit.
2017-11-10 23:14:30.0941 DEBUG Authenticated from cache.
2017-11-10 23:14:30.0981 DEBUG Session state was set to Stopped.
2017-11-10 23:14:30.6565 DEBUG Session state was set to Started.
2017-11-10 23:14:31.0189 DEBUG Sending RPC Request: 'GetPlayer, CheckChallenge'
2017-11-10 23:14:31.0229 DEBUG => Platform Request: 'SendEncryptedSignature, UnknownPtr8'
2017-11-10 23:14:31.6145 DEBUG Sending RPC Request: 'GetPlayer, CheckChallenge'
2017-11-10 23:14:31.6145 DEBUG => Platform Request: 'SendEncryptedSignature, UnknownPtr8'
2017-11-10 23:14:31.9333 DEBUG Received a new AuthTicket from Pokemon!
2017-11-10 23:14:33.0361 DEBUG Sending RPC Request: 'GetMapObjects, CheckChallenge, GetHatchedEggs, GetHoloInventory, CheckAwardedBadges, DownloadSettings, GetInbox, GetInbox'
2017-11-10 23:14:33.0361 DEBUG => Platform Request: 'SendEncryptedSignature, UnknownPtr8'
2017-11-10 23:14:33.2178  INFO Unknown status code: SessionInvalidated
2017-11-10 23:14:33.2448 ERROR SendRemoteProcedureCall exception: System.Exception: There were 0 responses.
   at POGOLib.Official.Net.RpcClient.HandleResponseEnvelope(RequestEnvelope requestEnvelope, ResponseEnvelope responseEnvelope) in L:\Gebruikers\Peter\Mijn documenten\Visual Studio 2017\Projects\PoGo.App\PogoLib\Furtif\POGOLib-core\POGOLib.Core\Net\RpcClient.cs:line 706
   at POGOLib.Official.Net.RpcClient.<PerformRemoteProcedureCallAsync>d__41.MoveNext() in L:\Gebruikers\Peter\Mijn documenten\Visual Studio 2017\Projects\PoGo.App\PogoLib\Furtif\POGOLib-core\POGOLib.Core\Net\RpcClient.cs:line 681
2017-11-10 23:14:33.3099 ERROR Map refresh failed: System.NullReferenceException: response
   at POGOLib.Official.Net.RpcClient.<RefreshMapObjectsAsync>d__33.MoveNext() in L:\Gebruikers\Peter\Mijn documenten\Visual Studio 2017\Projects\PoGo.App\PogoLib\Furtif\POGOLib-core\POGOLib.Core\Net\RpcClient.cs:line 310
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at POGOLib.Official.Pokemon.HeartbeatDispatcher.<Dispatch>d__7.MoveNext() in L:\Gebruikers\Peter\Mijn documenten\Visual Studio 2017\Projects\PoGo.App\PogoLib\Furtif\POGOLib-core\POGOLib.Core\Pokemon\HeartbeatDispatcher.cs:line 124
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at System.Runtime.CompilerServices.TaskAwaiter.GetResult()
   at POGOLib.Official.Pokemon.HeartbeatDispatcher.<CheckDispatch>d__4.MoveNext() in L:\Gebruikers\Peter\Mijn documenten\Visual Studio 2017\Projects\PoGo.App\PogoLib\Furtif\POGOLib-core\POGOLib.Core\Pokemon\HeartbeatDispatcher.cs:line 73
2017-11-10 23:14:33.3099  INFO No fort found nearby.

Furtif commented:
@PeterDM65 ok try again i m changed to old mode getmapobjects

Furtif committed:
fix https://github.com/AeonLucid/POGOLib/pull/97#issuecomment-343601678

PeterDM65 commented:
Works now ... thnx

PeterDM65 committed:
Fixed bug in calculating difference between current time and LastRpcRequest
https://github.com/Furtif/POGOLib/issues/2#issue-275158991

Furtif committed:
fix warning CS1998 and clean unsupported devices

Furtif committed:
update protos

PeterDM65 committed:
Bug fix on bug fix ... throttling cast to int at the wrong place

Furtif committed:
new api support

Furtif committed:
Fixes ptc login heads

Furtif committed:
Add PlayerLocal

Furtif committed:
Add Hash host and end point configuration for use other hash services

Furtif committed:
update, common requests have changed:
 - checkChallenge is not more
 - downloadSettings is always there now

Furtif committed:
check_challenge is back

Furtif committed:
0.85.1

Furtif committed:
try to run multi servers hash

Furtif committed:
fix multiserver hasher

Furtif committed:
add others player infos
